### PR TITLE
fix: httpInterceptor behavioural bug

### PR DESCRIPTION
**Pull request by Madhusudan092001**
## Clickup
app.clickup.com

## Code Coverage
Fixing the behavioural bug in httpInterceptor which occurred when we have multiple requests and the token is about to expire then triggering refreshToken some of the requests were getting old or null as token
Fixed by changing the intercept method and also using switchMap instead of concatMap

## UI Preview
Please add screenshots for UI changes

### CODE REVIEW

File: src/app/core/interceptors/httpInterceptor.ts [hunk 1]
```diff
@@ -8,27 +8,22 @@ import {
...
 import { UserEventService } from '../services/user-event.service';
-
 @Injectable()
 export class HttpConfigInterceptor implements HttpInterceptor {
-  public accessTokenCallInProgress = false;
-
-  public accessTokenSubject = new BehaviorSubject<string>(null);
-
+  private accessTokenCallInProgress = false;
+  // This BehaviorSubject coordinates access token updates across multiple requests
```
Comment (suyashpatil78) [inline src/app/core/interceptors/httpInterceptor.ts:25]: put a space before the comment

File: src/app/core/interceptors/httpInterceptor.ts [hunk 2]
```diff
@@ -51,147 +45,133 @@ export class HttpConfigInterceptor implements HttpInterceptor {
...
+      const differenceSeconds = expiryDate.diff(now, 'seconds');
       const maxRefreshDifferenceSeconds = 2 * 60;
       return differenceSeconds < maxRefreshDifferenceSeconds;
     } catch (err) {
       return true;
     }
   }
-
-  refreshAccessToken(): Observable<string> {
+  private refreshAccessToken(): Observable<string | null> {
```
Comment (suyashpatil78) [inline src/app/core/interceptors/httpInterceptor.ts:59]: put a space before the method

File: src/app/core/interceptors/httpInterceptor.ts [hunk 3]
```diff
@@ -51,147 +45,133 @@ export class HttpConfigInterceptor implements HttpInterceptor {
...
-        this.secureStorageService.clearAll();
-        this.storageService.clearAll();
-        globalCacheBusterNotifier.next();
-        return throwError(error);
-      }),
-      concatMap((authResponse) => this.routerAuthService.newAccessToken(authResponse.access_token)),
-      concatMap(() => from(this.tokenService.getAccessToken()))
+      switchMap((refreshToken) => {
+        if (refreshToken) {
+          // Fetch the new access token using the refresh token
```
Comment (suyashpatil78) [inline src/app/core/interceptors/httpInterceptor.ts:63]: such comments are not necessary as the method name is enough for knowing what this line do

File: src/app/core/interceptors/httpInterceptor.spec.ts [hunk 4]
```diff
@@ -14,7 +14,7 @@ import { BehaviorSubject, of, throwError } from 'rxjs';
 import { extendedDeviceInfoMockData, extendedDeviceInfoMockDataWoApp } from '../mock-data/extended-device-info.data';
 import { HttpErrorResponse, HttpHeaders, HttpRequest } from '@angular/common/http';
 
-describe('HttpConfigInterceptor', () => {
+fdescribe('HttpConfigInterceptor', () => {
```
Comment (suyashpatil78) [inline src/app/core/interceptors/httpInterceptor.spec.ts:17]: remove `fdescribe` and write `describe` instead. The code coverage will get fixed.

File: src/app/core/interceptors/httpInterceptor.ts [hunk 5]
```diff
@@ -55,130 +51,115 @@ export class HttpConfigInterceptor implements HttpInterceptor {
...
-            );
-          }
-        } else {
+      switchMap((accessToken) => {
+        if (accessToken && !this.expiringSoon(accessToken)) {
           return of(accessToken);
         }
+        if (!this.accessTokenCallInProgress) {
+          this.accessTokenCallInProgress = true;
+          this.accessTokenSubject.next(null);
```
Comment (Julias0) [inline src/app/core/interceptors/httpInterceptor.ts:103]: This line can be removed

File: src/app/fyle/personal-cards/personal-cards.page.ts [hunk 6]
```diff
@@ -471,7 +472,7 @@ export class PersonalCardsPage implements OnInit, AfterViewInit {
     if (this.selectionMode) {
       this.switchSelectionMode();
     }
-    this.selectedTrasactionType = event.detail.value as string;
+    this.selectedTrasactionType = event.detail.value;
```
Comment (suyashpatil78) [inline src/app/fyle/personal-cards/personal-cards.page.ts:475]: why are we changing this in the current PR?

Comment (Madhusudan092001) [reply src/app/fyle/personal-cards/personal-cards.page.ts:475]: I have created another PR  for some bug in personal-cards.page.ts , but for that i have fixed the lint error mistakenly here so this is one modification

**Review by Julias0 — APPROVED** [review]

**Pull request merged**